File: apps/java/src/main/java/com/example/demo/auth/AuthService.java
package com.example.demo.auth;

import com.example.demo.config.SessionProperties;
import com.example.demo.exercises.db.User;
import com.example.demo.exercises.db.UserRepository;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

  private final UserRepository users;
  private final RedisSessionRepository sessions;
  private final SessionProperties sessionProperties;
  private final PasswordHasher passwordHasher;

  public AuthService(
      UserRepository users,
      RedisSessionRepository sessions,
      SessionProperties sessionProperties,
      PasswordHasher passwordHasher) {
    this.users = users;
    this.sessions = sessions;
    this.sessionProperties = sessionProperties;
    this.passwordHasher = passwordHasher;
  }

  public EnsureSessionResult ensureSession(String clientSessionId, SharedSession requestSession) {
    Instant now = Instant.now();
    Optional<SharedSession> fromRequest = validSession(requestSession, now);
    if (fromRequest.isPresent()) {
      return new EnsureSessionResult(fromRequest.get(), false);
    }
    if (clientSessionId != null && !clientSessionId.isBlank()) {
      Optional<SharedSession> fromClient = validStoredSession(clientSessionId.trim(), now);
      if (fromClient.isPresent()) {
        return new EnsureSessionResult(fromClient.get(), false);
      }
    }
    return new EnsureSessionResult(createAnonymousSession(), true);
  }

  public SharedSession login(LoginRequest request) {
    User user = resolveUser(request);
    verifyPassword(request, user);
    // New Redis key for this login; leave any prior session (e.g. guest in other tabs) untouched.
    Instant issuedAt = Instant.now();
    Instant expiresAt = issuedAt.plus(sessionProperties.getTtl());
    SharedSession session =
        new SharedSession(
            UUID.randomUUID().toString(),
            user.getId(),
            user.getEmail(),
            user.getName(),
            issuedAt,
            expiresAt,
            "java");
    sessions.save(session);
    return session;
  }

  public void logout(String sessionId) {
    sessions.delete(sessionId);
  }

  /** Delete the cookie session id (if any) and issue a fresh guest session in Redis. */
  public SharedSession logoutAndCreateGuest(String sessionId) {
    if (sessionId != null && !sessionId.isBlank()) {
      sessions.delete(sessionId.trim());
    }
    return createAnonymousSession();
  }

  /** Delete the current Redis session (if any) and issue a new session id with fresh payload. */
  public SharedSession refreshSession(SharedSession current) {
    if (current != null) {
      sessions.delete(current.sessionId());
    }
    Instant issuedAt = Instant.now();
    Instant expiresAt = issuedAt.plus(sessionProperties.getTtl());
    SharedSession session;
    if (current != null && current.userId() > 0L) {
      session =
          new SharedSession(
              UUID.randomUUID().toString(),
              current.userId(),
              current.email(),
              current.name(),
              issuedAt,
              expiresAt,
              "java");
    } else {
      session =
          new SharedSession(
              UUID.randomUUID().toString(),
              0L,
              null,
              "Guest",
              issuedAt,
              expiresAt,
              "java");
    }
    sessions.save(session);
    return session;
  }

  public SharedSession requireCurrentSession() {
    SharedSession session = SessionContext.get();
    if (session == null) {
      throw new AuthException(HttpStatus.UNAUTHORIZED, "No active session");
    }
    if (session.isExpired(Instant.now())) {
      sessions.delete(session.sessionId());
      throw new AuthException(HttpStatus.UNAUTHORIZED, "Session expired");
    }
    return session;
  }

  public SharedSession validateSessionId(String sessionId) {
    SharedSession session =
        sessions
            .findById(sessionId)
            .orElseThrow(
                () -> new AuthException(HttpStatus.UNAUTHORIZED, "Session not found in Redis"));
    if (session.isExpired(Instant.now())) {
      sessions.delete(session.sessionId());
      throw new AuthException(HttpStatus.UNAUTHORIZED, "Session expired");
    }
    return session;
  }

  public String redisKey(String sessionId) {
    return sessionProperties.redisKey(sessionId);
  }

  private SharedSession createAnonymousSession() {
    Instant issuedAt = Instant.now();
    Instant expiresAt = issuedAt.plus(sessionProperties.getTtl());
    SharedSession session =
        new SharedSession(
            UUID.randomUUID().toString(),
            0L,
            null,
            "Guest",
            issuedAt,
            expiresAt,
            "java");
    sessions.save(session);
    return session;
  }

  private static Optional<SharedSession> validSession(SharedSession session, Instant now) {
    if (session == null || session.isExpired(now)) {
      return Optional.empty();
    }
    return Optional.of(session);
  }

  private Optional<SharedSession> validStoredSession(String sessionId, Instant now) {
    return sessions.findById(sessionId).filter(session -> !session.isExpired(now));
  }

  private void verifyPassword(LoginRequest request, User user) {
    if (request.password() == null || request.password().isBlank()) {
      return;
    }
    String hash = user.getPasswordHash();
    if (hash == null || hash.isBlank()) {
      throw new AuthException(
          HttpStatus.UNAUTHORIZED, "Password login is not available for this user");
    }
    if (!passwordHasher.matches(request.password(), hash)) {
      throw new AuthException(HttpStatus.UNAUTHORIZED, "Invalid email or password");
    }
  }

  private User resolveUser(LoginRequest request) {
    if (request == null) {
      throw new AuthException(HttpStatus.BAD_REQUEST, "email or userId is required");
    }
    if (request.email() != null && !request.email().isBlank()) {
      return users
          .findByEmailIgnoreCase(request.email().trim())
          .orElseThrow(
              () ->
                  new AuthException(
                      HttpStatus.NOT_FOUND, "No user with email " + request.email().trim()));
    }
    if (request.userId() != null) {
      return users
          .findById(request.userId())
          .orElseThrow(
              () ->
                  new AuthException(
                      HttpStatus.NOT_FOUND, "No user with id " + request.userId()));
    }
    throw new AuthException(HttpStatus.BAD_REQUEST, "email or userId is required");
  }
}

File: apps/python/src/exercises/web/auth_service.py
"""Session ensure/login/logout (mirrors Java AuthService / Rust auth service)."""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import timedelta

from exercises.web.db import DatabaseNotConfiguredError, connection, find_user_auth_by_email, find_user_by_email, find_user_by_id
from exercises.web.password_util import verify_password
from exercises.web.session_models import SessionConfig, SharedSession, utc_now
from exercises.web.session_repository import SessionRepository


@dataclass
class EnsureSessionResult:
    session: SharedSession
    created: bool


class AuthServiceError(Exception):
    def __init__(self, status: int, message: str) -> None:
        super().__init__(message)
        self.status = status
        self.message = message


def ensure_session(
    repo: SessionRepository,
    config: SessionConfig,
    client_session_id: str | None,
    request_session: SharedSession | None,
) -> EnsureSessionResult:
    now = utc_now()
    if request_session is not None and not request_session.is_expired(now):
        return EnsureSessionResult(request_session, created=False)
    if client_session_id and client_session_id.strip():
        stored = repo.find_by_id(client_session_id.strip())
        if stored is not None and not stored.is_expired(now):
            return EnsureSessionResult(stored, created=False)
        if stored is not None:
            repo.delete(stored.session_id)
    session = _create_anonymous_session(repo, config)
    return EnsureSessionResult(session, created=True)


def login(
    repo: SessionRepository,
    config: SessionConfig,
    *,
    email: str | None,
    user_id: int | None,
    password: str | None = None,
    request_id: str | None = None,
) -> SharedSession:
    user = _resolve_user(email=email, user_id=user_id, password=password, request_id=request_id)
    issued_at = utc_now()
    expires_at = issued_at + timedelta(seconds=config.ttl_secs)
    session = SharedSession(
        session_id=str(uuid.uuid4()),
        user_id=user[0],
        email=user[2],
        name=user[1],
        issued_at=issued_at,
        expires_at=expires_at,
        issuer="python",
    )
    repo.save(session)
    return session


def logout(repo: SessionRepository, session_id: str) -> None:
    repo.delete(session_id)


def logout_and_create_guest(
    repo: SessionRepository,
    config: SessionConfig,
    session_id: str | None,
) -> SharedSession:
    """Delete the cookie session id (if any) and issue a fresh guest session in Redis."""
    if session_id and session_id.strip():
        repo.delete(session_id.strip())
    return _create_anonymous_session(repo, config)


def refresh_session(
    repo: SessionRepository,
    config: SessionConfig,
    current: SharedSession | None,
) -> SharedSession:
    """Delete the current Redis session (if any) and issue a new session id."""
    if current is not None:
        repo.delete(current.session_id)
    issued_at = utc_now()
    expires_at = issued_at + timedelta(seconds=config.ttl_secs)
    if current is not None and current.user_id > 0:
        session = SharedSession(
            session_id=str(uuid.uuid4()),
            user_id=current.user_id,
            email=current.email,
            name=current.name,
            issued_at=issued_at,
            expires_at=expires_at,
            issuer="python",
        )
    else:
        session = SharedSession(
            session_id=str(uuid.uuid4()),
            user_id=0,
            email=None,
            name="Guest",
            issued_at=issued_at,
            expires_at=expires_at,
            issuer="python",
        )
    repo.save(session)
    return session


def _create_anonymous_session(repo: SessionRepository, config: SessionConfig) -> SharedSession:
    issued_at = utc_now()
    expires_at = issued_at + timedelta(seconds=config.ttl_secs)
    session = SharedSession(
        session_id=str(uuid.uuid4()),
        user_id=0,
        email=None,
        name="Guest",
        issued_at=issued_at,
        expires_at=expires_at,
        issuer="python",
    )
    repo.save(session)
    return session


def _resolve_user(
    *,
    email: str | None,
    user_id: int | None,
    password: str | None = None,
    request_id: str | None,
) -> tuple[int, str, str]:
    trimmed_email = email.strip() if email else ""
    if trimmed_email:
        try:
            with connection(request_id=request_id) as conn:
                auth_row = find_user_auth_by_email(conn, trimmed_email)
        except DatabaseNotConfiguredError as exc:
            raise AuthServiceError(503, str(exc)) from exc
        if auth_row is None:
            raise AuthServiceError(404, f"No user with email {trimmed_email}")
        if password:
            if not auth_row[3] or not verify_password(password, auth_row[3]):
                raise AuthServiceError(401, "Invalid email or password")
        return auth_row[0], auth_row[1], auth_row[2]
    if user_id is not None:
        try:
            with connection(request_id=request_id) as conn:
                row = find_user_by_id(conn, user_id)
        except DatabaseNotConfiguredError as exc:
            raise AuthServiceError(503, str(exc)) from exc
        if row is None:
            raise AuthServiceError(404, f"No user with id {user_id}")
        return row
    raise AuthServiceError(400, "email or userId is required")
